pkg/odo/genericclioptions/context.go

```diff
@@ -0,0 +1,204 @@
+		util.CheckError(err, "could not get current component")
+	} else {
+		// if flag is set, check that the specified component exists
+		context.existsOrExit(cmpFlag)
+		cmp = cmpFlag
+	}
+
+	// once the component is resolved, add it to the context
+	context.cmp = cmp
+
+	return context
+}
+
+// FlagValueIfSet retrieves the value of the specified flag if it is set for the given command
+func FlagValueIfSet(cmd *cobra.Command, flagName string) string {
+	flag, err := cmd.Flags().GetString(flagName)
+
+	// log the error for debugging purposes though an error should only occur if the flag hadn't been added to the command or
+	// if the specified flag name doesn't match a string flag. This usually can be ignored.
+	ignoreButLog(err)
+	return flag
+}
+
+// Context holds contextual information useful to commands such as correctly configured client, target project and application
+// (based on specified flag values) and provides for a way to retrieve a given component given this context
+type Context struct {
+	internalCxt
+}
+
+// internalCxt holds the actual context values and is not exported so that it cannot be instantiated outside of this package.
+// This ensures that Context objects are always created properly via NewContext factory functions.
+type internalCxt struct {
+	Client      *occlient.Client
+	Project     string
+	Application string
+	cmp         string
+}
+
+// Component retrieves the optionally specified component or the current one if it is set. If no component is set, exit with
+// an error
+func (o *Context) Component(optionalComponent ...string) string {
+	return o.ComponentAllowingEmpty(false, optionalComponent...)
+}
+
+// ComponentAllowingEmpty retrieves the optionally specified component or the current one if it is set, allowing empty
+// components (instead of exiting with an error) if so specified
+func (o *Context) ComponentAllowingEmpty(allowEmpty bool, optionalComponent ...string) string {
+	switch len(optionalComponent) {
+	case 0:
+		// if we're not specifying a component to resolve, get the current one (resolved in NewContext as cmp)
+		// so nothing to do here unless the calling context doesn't allow no component to be set in which case we exit with error
+		if !allowEmpty && len(o.cmp) == 0 {
+			fmt.Println("No component is set")
+			os.Exit(-1)
+		}
+	case 1:
+		cmp := optionalComponent[0]
+		// only check the component if we passed a non-empty string, otherwise return the current component set in NewContext
+		if len(cmp) > 0 {
+			o.existsOrExit(cmp)
+			o.cmp = cmp // update context
+		}
+	default:
+		// safeguard: fail if more than one optional string is passed because it would be a programming error
+		fmt.Printf("ComponentAllowingEmpty function only accepts one optional argument, was given: %v", optionalComponent)
+		os.Exit(-1)
+	}
+
+	return o.cmp
+}
+
+// existsOrExit checks if the specified component exists with the given context and exits the app if not.
+func (o *Context) existsOrExit(cmp string) {
```

Comment:
checkComponentExistsOrFail ?

Reply:
I thought that `existsOrExit` was pretty explicit but I can change it if needed.

Reply:
Changed!